tracer/test/Datadog.Trace.Security.Unit.Tests/ApiSecurityTests.cs

```diff
@@ -0,0 +1,151 @@
+        }
+
+        res.Should().Be(expectedResult);
+        if (expectedResult)
+        {
+            dic.Should().ContainKey(AddressesConstants.WafContextProcessor);
+        }
+    }
+
+    [Fact]
+    public void ApiSecurityTestMaxRoutes()
+    {
+        var maxRouteSize = 50;
+        var apiSec = new ApiSecurity(new SecuritySettings(new CustomSettingsForTests(new Dictionary<string, object> { { Configuration.ConfigurationKeys.AppSec.ApiExperimentalSecurityEnabled, true } }), new NullConfigurationTelemetry()), maxRouteSize);
+        var queue = new Queue<int>(maxRouteSize);
+        for (var i = 0; i < maxRouteSize + 1; i++)
+        {
+            var dt = DateTime.UtcNow;
+            if (queue.Count == maxRouteSize)
+            {
+                queue.Dequeue();
+            }
+
+            var route = $"route{i}";
+            var method = $"GET{i}";
+            var statusCode = i.ToString();
+            var resHash = ApiSecurity.CombineHashes(route, method, statusCode);
+            queue.Enqueue(resHash);
+            var dic = new Dictionary<string, object>();
+            var tc = new TraceContext(Mock.Of<IDatadogTracer>(), new TraceTagCollection());
+            tc.SetSamplingPriority(SamplingPriorityValues.AutoKeep);
+
+            var span = new Span(new SpanContext(SpanContext.None, tc, "Test"), dt);
+            span.SetTag(Tags.HttpRoute, route);
+            span.SetTag(Tags.HttpStatusCode, statusCode);
+            span.SetTag(Tags.HttpMethod, method);
+            var res = apiSec.ShouldAnalyzeSchema(true, span, dic, statusCode, new Dictionary<string, object>());
+            res.Should().BeTrue();
+        }
+
+        var innerDic = GetInnerOrderedDictionary(apiSec);
+        innerDic.Count.Should().Be(maxRouteSize);
+        var oldestElm = queue.First();
+        var oldest = innerDic.Keys.Cast<int>().First();
+        oldest.Should().Be(oldestElm);
+    }
+
+    [Fact]
+    public void ApiSecurityTestMultiThread()
+    {
+        var apiSec = new ApiSecurity(
+            new SecuritySettings(
+                new CustomSettingsForTests(
+                    new Dictionary<string, object> { { Configuration.ConfigurationKeys.AppSec.ApiExperimentalSecurityEnabled, true }, { Configuration.ConfigurationKeys.AppSec.ApiSecuritySampleDelay, 120 } }),
+                new NullConfigurationTelemetry()));
+        var dic = new Dictionary<string, object> { { "controller", "test" }, { "action", "test" } };
+        var tc = new TraceContext(Mock.Of<IDatadogTracer>(), new TraceTagCollection());
+        tc.SetSamplingPriority(SamplingPriorityValues.AutoKeep);
+        var dt = DateTime.UtcNow;
+
+        var span = new Span(new SpanContext(SpanContext.None, tc, "Test"), dt);
+        span.SetTag(Tags.HttpRoute, "{controller}/{action}");
+        span.SetTag(Tags.HttpStatusCode, "200");
+        span.SetTag(Tags.HttpMethod, "GET");
+        List<Thread> threads = new();
+        ConcurrentBag<bool> results = new();
```

Comment:
[Never use ConcurrentBag](https://minidump.net/performance-best-practices-in-c-b85a47bdd93a/#concurrentbagt) according to @kevingosse  😄

Reply:
but within a test? 😇

Reply:
changed to `ConcurrentQueue `after reading the article 💯  in [8d4eab8](https://github.com/DataDog/dd-trace-dotnet/pull/5257/commits/8d4eab8a5a89b0b295060aa769fa8e9101c3bff8)

Reply:
Oops, didn't spot it was in a test, I've just been trained by Kevin to be triggered by it 😂